src/Compilers/Core/Portable/Operations/Operation.cs

```diff
@@ -46,41 +83,203 @@ public Operation(OperationKind kind, SyntaxNode syntax, ITypeSymbol type, Option
 
         public abstract TResult Accept<TArgument, TResult>(OperationVisitor<TArgument, TResult> visitor, TArgument argument);
 
-        public static IOperation CreateOperationNone(SyntaxNode node, Optional<object> constantValue, Func<ImmutableArray<IOperation>> getChildren)
+        protected void SetParentOperation(IOperation parent)
         {
-            return new NoneOperation(node, constantValue, getChildren);
+            var result = Interlocked.CompareExchange(ref _parentDoNotAccessDirectly, parent, null);
+#if DEBUG
+            if (result == null)
+            {
+                // tree must belong to same semantic model
+                Debug.Assert(((Operation)parent).SemanticModel == SemanticModel);
+            }
+            else
+            {
+                // make sure given parent and one we already have is same
+                Debug.Assert(result == parent);
+            }
+#endif
         }
 
-        private class NoneOperation : IOperation
+        public static IOperation CreateOperationNone(SemanticModel semanticModel, SyntaxNode node, Optional<object> constantValue, Func<ImmutableArray<IOperation>> getChildren)
         {
-            private readonly Func<ImmutableArray<IOperation>> _getChildren;
+            return new NoneOperation(semanticModel, node, constantValue, getChildren);
+        }
 
-            public NoneOperation(SyntaxNode node, Optional<object> constantValue, Func<ImmutableArray<IOperation>> getChildren)
+        public static T SetParentOperation<T>(T operation, IOperation parent) where T : IOperation
```

Comment:
IOperation has parent and child pointer, one of them has to be set outside of constructor. I choose parent since it didn't require changing all factory code which is hand written rather than generated one.